Smart linebreak printer for pipe chains

fhammerschmidt opened the pull request:
This PR is an alternative to #6387
Fixes #6380

glennsl commented on jscomp/syntax/tests/printer/expr/expected/callback.res.txt:
```diff
@@ -215,13 +215,16 @@ let f = () => {
 }
 
-myPromise->Js.Promise.then_(value => {
+myPromise
+->Js.Promise.then_(value => {
```
This is perhaps a bit unfortunate. Not so bad with multiple pipes, but does it also break with just one pipe if the function call is over multiple lines? E.g.:

```rescript
myPromise->Promise.then(v => {
  Js.log(v)
  Promise.resolve(v)
})


// vs

myPromise
->Promise.then(v => {
  Js.log(v)
  Promise.resolve(v)
})
```

fhammerschmidt replied:
I will try to fix this edge case. Currently the latter will be formatted to the former. Thank you for the fast feedback.

glennsl replied:
Yes, the former is what I'd prefer. It would be good to have a test for this, unless there already is one and I missed it.

fhammerschmidt replied:
Just added your example to the tests.

glennsl replied:
Awesome. Thank you!

aspeddro commented on jscomp/syntax/tests/conversion/reason/fastPipe.res:
```diff
@@ -22,3 +22,13 @@ let aggregateTotal = (forecast, ~audienceType) =>
+"expr1"
+->myFunc("expr2")->myFunc("expr3")
+
+"expr1"->myFunc("expr2")
+->myFunc("expr3")->myFunc("expr4")
```
I found this very confusing

This is cleaner

```suggestion
"expr1"
->myFunc("expr2")
->myFunc("expr3")
->myFunc("expr4")
```

aspeddro commented on jscomp/syntax/tests/conversion/reason/expected/fastPipe.res.txt:
```diff
@@ -22,3 +22,16 @@ let aggregateTotal = (forecast, ~audienceType) =>
+
+let myFunc = (strA, strB) => strA ++ strB
+
+"expr1"->myFunc("expr2")->myFunc("expr3")
```
Why doesn't this break into several lines?

fhammerschmidt replied:
Because it has  no newline in between the expressions

fhammerschmidt commented:
Found out that it did not break correctly if the newline was before the last pipe (no matter how many pipes there were in total). Fixed now.
Also moved all smart pipe tests to a separate file.
Also added a formatting test with a React Hook which is now more densely formatted (the same as when there was a {} block in it).

fhammerschmidt commented:
I formatted a big codebase with this change and the only thing that changed were some Hooks like this one: 

```diff
-  React.useCallback(
-    () =>
-      flatListRef.current
-      ->Nullable.toOption
-      ->Option.forEach(flatList => flatList->FlatList.scrollToOffset({offset: 0.})),
-    [],
-  )
+  React.useCallback(() =>
+    flatListRef.current
+    ->Nullable.toOption
+    ->Option.forEach(flatList => flatList->FlatList.scrollToOffset({offset: 0.}))
+  , [])
```

However, if you wrap the former callback body in curly braces it was already formatted like that:

```res
  React.useCallback(() => {
    flatListRef.current
    ->Nullable.toOption
    ->Option.forEach(flatList => flatList->FlatList.scrollToOffset({offset: 0.}))
  }, [])
```

If this change is ok, the PR is ready to review
@zth @cristianoc @cknitt

zth commented:
Looks great to me! @cristianoc ?

cristianoc commented:
> Looks great to me! @cristianoc ?

Looks great

zth commented:
@fhammerschmidt fix that merge conflict and then let's merge? This is really worthy of a forum post as well given this is something many have complained about over the years. Great work! ⭐
